src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe_expert_opt.cpp

```diff
@@ -0,0 +1,1215 @@
+                                            dnnl_weights[2].zp);
+        _kernels[key] = kernel;
+        return _kernels[key];
+    }
+
+    //  inputs 0 is hidden_states, inputs 1 is router_logits[num_tokens, NUM_EXPERTS=128]
+    //  extra step Softmax_TopK is fused to give topk-id & router_weights
+    //
+    //     scratch.topk_id, scratch.full_router_weights = Softmax_TopK(router_logits)
+    //
+    //  generate expert_mask from topk-id
+    //        expert_mask.batch[i][j] : j'th token index for i'th expert
+    //        expert_mask.topk[i][j] : topk-output offset for j'th token for i'th expert, used to get weights
+    //        expert_mask.pred_flag[i]: bool, if expert i can be skipped
+    //
+    //
+    //     scratch.x, scratch.routing_weights = gather(hidden_states, scratch.full_router_weights, expert_mask.batch, expert_mask.topk)
+    //     scratch.y = MLP(scratch.x, .gate/up/down) * scratch.routing_weights
+    //     scatter(final_hidden, scratch.y, expert_mask.batch)
+    //
+    cldnn::event::ptr execute(const std::vector<cldnn::event::ptr>& events, cldnn::primitive_inst& ins) override {
+        OV_ITT_SCOPED_TASK(ov::intel_gpu::itt::domains::intel_gpu_plugin, openvino::itt::handle("MoeExpertOptImpl::execute"));
+        auto& instance = reinterpret_cast<typed_primitive_inst<moe_expert>&>(ins);
+        auto moe = instance.get_typed_desc<moe_expert>();
+        const auto& config = moe->_config;
+        int max_topk = static_cast<int>(config.topk);
+        auto& cur_net = instance.get_network();
+        auto& stream = cur_net.get_stream();
+
+        auto [hidden_states_mem_ptr, hidden_states_layout] = get_input_info(instance, 0);
+        auto batch = static_cast<int>(hidden_states_layout.get_shape()[0]);
+
+        scratch_buffers scratch;
+        prepare_internal_buffers(instance, scratch, batch == 1);
+
+        // softmax+topk
+        auto lws_size = moe->_config.expert_num;
+        execute_stage(events,
+                      instance,
+                      *softmax_topk,
+                      {instance.input_memory_ptr(1)},
+                      {scratch.topk_id, scratch.topk_weights},
+                      {static_cast<size_t>(batch), lws_size},
+                      {1, lws_size});
+
+        // Single batch is a special case, we don't need to do gather/scatter,
+        // and we can apply optimal kernels against memory bound to improve performance.
+        // It is very important for MoE's second token performance.
+        if (batch == 1) {
+            return exec_single_batch(instance, scratch);
+        }
+
+        auto final_hidden_states_mem_ptr = instance.output_memory_ptr(0);
+        auto final_hidden_states_layout = instance.get_output_layout(0);
+
+        // onednn path will accumulate to the output
+        final_hidden_states_mem_ptr->fill(stream, false);
+
+        expert_mask_cpu expert_mask;
+        // Wait for topk is ready
+        stream.finish();
```

Comment:
It would be better to use softmax_topk output event for this synchronization to reduce finish() calls number

Reply:
updated